### PR TITLE
Create process to build proto sources

### PR DESCRIPTION
**Pull request by rsgowman**
This PR allows building of the proto sources. The old script has suffered bit rot, and no longer worked. 

The new script is not yet hooked into the travis build. (TODO)

As part of this, we now only support generating protos via the cmake build, so it's no longer possible to use the podfile based method. (The nanopb pod doesn't support generating nanopb sources from proto files.) Previously, the build script required both build methods.

A number of cleanups were also performed.

### CODE REVIEW

**Comment by rsgowman** [comment]
I've split this from the PR that integrates this into travis for no reason other than to get early feedback.

This PR makes the situation "no worse", so I'm comfortable submitting with that, but we can also wait for travis integration if preferred.

File: Firestore/Protos/CMakeLists.txt [hunk 1]
```diff
@@ -97,3 +97,57 @@ target_include_directories(
   firebase_firestore_protos_libprotobuf PUBLIC
   ${FIREBASE_SOURCE_DIR}/Firestore/Protos/cpp
 )
+
+set(NANOPB_GENERATOR_DIR ${FIREBASE_BINARY_DIR}/external/src/nanopb/generator/)
```
Comment (wilhuff) [inline Firestore/Protos/CMakeLists.txt]: trailing `/`

File: Firestore/Protos/CMakeLists.txt [hunk 2]
```diff
@@ -97,3 +97,57 @@ target_include_directories(
...
+
+# Build nanopb protoc plugin. (Heavily "inspired" by
+# https://github.com/nanopb/nanopb/blob/master/extra/FindNanopb.cmake)
+set(NANOPB_GENERATOR_SRC
+  ${NANOPB_GENERATOR_DIR}/proto/nanopb.proto
+  ${NANOPB_GENERATOR_DIR}/proto/plugin.proto
+)
+set(NANOPB_GENERATOR_PYTHON_SRC)
+foreach (FIL ${NANOPB_GENERATOR_SRC})
```
Comment (wilhuff) [inline Firestore/Protos/CMakeLists.txt]: This loop still knows about `${NANOPB_GENERATOR_DIR}/proto`. Why not make `NANOPB_GENERATOR_SRC` be `nanopb plugin` and then get rid of all the `get_filename_component` abstraction in here?

```cmake
foreach (GENERATOR ${NANOPB_GENERATOR_SRC})
  set(output "${NANOPB_GENERATOR_DIR}/proto/${GENERATOR}_pb2.py")
  # ...
endforeach()
```

Possibly also rename to `NANOPB_GENERATORS` or _BASENAMES or something like that too.

File: Firestore/Protos/nanopb/firestore/local/maybe_document.nanopb.c [hunk 3]
```diff
@@ -15,7 +15,7 @@
  */
 
 /* Automatically generated nanopb constant definitions */
-/* Generated by nanopb-0.3.8 at Fri Jun  1 18:36:44 2018. */
+/* Generated by nanopb-0.3.9.1 at Tue Aug 14 12:19:56 2018. */
```
Comment (wilhuff) [inline Firestore/Protos/nanopb/firestore/local/maybe_document.nanopb.c]: Is this going to work with the nanopb [0.3.8 that's available via CocoaPods](https://cocoapods.org/pods/nanopb)?

File: Firestore/Protos/CMakeLists.txt (hunk 1 above)
Comment (rsgowman) [reply Firestore/Protos/CMakeLists.txt]: Done.

File: Firestore/Protos/CMakeLists.txt (hunk 2 above)
Comment (rsgowman) [reply Firestore/Protos/CMakeLists.txt]: Short answer: 'cause I blindly copied from nanopb.

Slightly longer answer: nanopb uses these values twice. In once case, it needs the full path name, in the other, only the components. So nanopb chose to store the fullname and extract where necessary. (They could've equivalently stored the components, and built the full path where necessary.)

We only need the values once though. So I've switched it as you suggested.

File: Firestore/Protos/nanopb/firestore/local/maybe_document.nanopb.c (hunk 3 above)
Comment (rsgowman) [reply Firestore/Protos/nanopb/firestore/local/maybe_document.nanopb.c]: It should. Note the check on line 23. That's what nanopb uses to determine if files need to be rebuilt. This assumes nanopb doesn't introduce incompatible changes without incrementing that number.

**Pull request merged**